pkg/controller/compute/firewall.go

```diff
@@ -0,0 +1,184 @@
+	kube client.Client
+}
+
+func (c *firewallConnector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	projectID, opts, err := gcp.GetAuthInfo(ctx, c.kube, mg)
+	if err != nil {
+		return nil, err
+	}
+	s, err := compute.NewService(ctx, opts)
+	if err != nil {
+		return nil, errors.Wrap(err, errNewClient)
+	}
+	return &firewallExternal{Service: s, kube: c.kube, projectID: projectID}, nil
+}
+
+type firewallExternal struct {
+	kube client.Client
+	*compute.Service
+	projectID string
+}
+
+func (c *firewallExternal) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.Firewall)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotFirewall)
+	}
+	observed, err := c.Firewalls.Get(c.projectID, meta.GetExternalName(cr)).Context(ctx).Do()
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(gcp.IsErrorNotFound, err), errGetFirewall)
+	}
+
+	currentSpec := cr.Spec.ForProvider.DeepCopy()
+	firewall.LateInitializeSpec(&cr.Spec.ForProvider, *observed)
+	if !cmp.Equal(currentSpec, &cr.Spec.ForProvider) {
+		if err := c.kube.Update(ctx, cr); err != nil {
+			return managed.ExternalObservation{}, errors.Wrap(err, errManagedFirewallUpdate)
+		}
+	}
+
+	cr.Status.AtProvider = firewall.GenerateFirewallObservation(*observed)
+
+	cr.Status.SetConditions(xpv1.Available())
+
+	u, err := firewall.IsUpToDate(meta.GetExternalName(cr), &cr.Spec.ForProvider, observed)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errCheckFirewallUpToDate)
+	}
+
+	return managed.ExternalObservation{
+		ResourceExists:   true,
+		ResourceUpToDate: u,
+	}, nil
+}
+
+func (c *firewallExternal) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1alpha1.Firewall)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotFirewall)
+	}
+
+	cr.Status.SetConditions(xpv1.Creating())
+
+	fw := &compute.Firewall{}
+	firewall.GenerateFirewall(meta.GetExternalName(cr), cr.Spec.ForProvider, fw)
+	_, err := c.Firewalls.Insert(c.projectID, fw).
+		Context(ctx).
+		Do()
+	return managed.ExternalCreation{}, errors.Wrap(err, errFirewallCreateFailed)
+}
+
+func (c *firewallExternal) Update(ctx context.Context, mg resource.Managed) (managed.ExternalUpdate, error) {
+	cr, ok := mg.(*v1alpha1.Firewall)
+	if !ok {
+		return managed.ExternalUpdate{}, errors.New(errNotFirewall)
+	}
+
+	observed, err := c.Firewalls.Get(c.projectID, meta.GetExternalName(cr)).Context(ctx).Do()
+	if err != nil {
+		return managed.ExternalUpdate{}, errors.Wrap(resource.Ignore(gcp.IsErrorNotFound, err), errGetFirewall)
+	}
+
+	upToDate, err := firewall.IsUpToDate(meta.GetExternalName(cr), &cr.Spec.ForProvider, observed)
+	if err != nil {
+		return managed.ExternalUpdate{}, errors.Wrap(err, errCheckFirewallUpToDate)
+	}
+	if upToDate {
+		return managed.ExternalUpdate{}, nil
+	}
```

Comment:
I believe we already did our observation in `Observe` above. Any reason to make another observation here?